src/item.cpp

```diff
@@ -5983,55 +5982,111 @@ bool item::allow_crafting_component() const
+    float new_item_temperature;
+
+    if( thermal_energy > completely_liquid_energy ) {
+        // Item is liquid
+        new_item_temperature = freezing_temperature + ( thermal_energy - completely_liquid_energy ) /
+                               ( specific_heat_liquid * mass );
+    } else if( thermal_energy < completely_frozen_energy ) {
+        // Item is solid
+        new_item_temperature = thermal_energy / ( specific_heat_solid * mass );
+    } else {
+        // Item is partially solid
+        new_item_temperature = freezing_temperature;
     }
-    return counters;
+    return new_item_temperature;
 }
 
-void item::set_temp_from_static( const int counters )
+float item::get_energy_from_temperature( const float new_temperature )
 {
-    item_tags.erase( "COLD" );
-    item_tags.erase( "HOT" );
+    const int specific_heat_liquid = type->comestible->spec_heat_liquid; // J/g K
+    const int specific_heat_solid = type->comestible->spec_heat_solid; // J/g K
+    const int latent_heat = type->comestible->lat_heat; // J/kg
+    const float mass = to_gram( weight() ); // g
+    const float freezing_temperature = temp_to_kelvin( type->comestible->freeze_point );  // K
+    const float completely_frozen_energy = mass * specific_heat_solid *
+                                           freezing_temperature;  // Energy that the item would have if it was completely solid at freezing temperature
+    const float completely_liquid_energy = completely_frozen_energy + mass *
+                                           latent_heat; // Energy that the item would have if it was completely liquid at freezing temperature
+    float thermal_energy;
 
-    if( counters < -1200 ) {
-        item_tags.insert( "FROZEN" );
-        item_counter = -counters - 1200;
-        if( current_phase == LIQUID ) {
-            current_phase = SOLID;
+    if( new_temperature <= freezing_temperature ) {
+        thermal_energy = completely_frozen_energy - mass * specific_heat_solid *
+                         ( freezing_temperature - new_temperature );
+    } else {
+        thermal_energy = completely_liquid_energy + mass * specific_heat_liquid *
+                         ( new_temperature - freezing_temperature );
+    }
+    return thermal_energy;
+}
+
+void item::set_temperature( float new_temperature )
+{
+    const float freezing_temperature = temp_to_kelvin( type->comestible->freeze_point );  // K
+    const float specific_heat_solid = type->comestible->spec_heat_solid; // J/g K
+    const float latent_heat = type->comestible->lat_heat; // J/kg
+    const float mass = to_gram( weight() ); // g
+
+    float new_specific_energy = get_energy_from_temperature( new_temperature );
+    float freeze_percentage = 0;
+
+    temperature = static_cast<int>( 100000 * new_temperature + 0.5 );
+    specific_energy = static_cast<int>( 100000 * new_specific_energy / mass + 0.5 );
```

Comment:
Every time ``specific_energy`` is used it is multiplied by 0.00001 or checked if it is negative. Why not to use integer ``specific_energy`` everywhere instead of ``new_specific_energy``? An make its value calculated in 10mkJ/g instead of J/g for higher precision.
Same for ``temperature`` and ``new_temperature``.

Reply:
Negative specific energy is just the default for new items. This way new items get immediately processed instead of waiting for 10 minutes.

Either the `specific_energy` is multiplied by 0.00001 or the `specific_heat_liquid`, `specific_heat_solid` and `latent_heat` are multiplied by 100000. These need to be in same units.

But since the temperature is also tracked in 0.00001 kelvins the numbers would maybe cancel out.